pkg/operator/controller/externaldns/deployment.go

```diff
@@ -0,0 +1,274 @@
+		ImagePullPolicy:          corev1.PullIfNotPresent,
+		TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+		Args:                     args,
+	}, nil
+}
+
+// createExternalDNSDeployment creates the given deployment using the reconciler's client.
+func (r *reconciler) createExternalDNSDeployment(ctx context.Context, depl *appsv1.Deployment) error {
+	if err := r.client.Create(ctx, depl); err != nil {
+		return fmt.Errorf("failed to create externalDNS deployment %s/%s: %w", depl.Namespace, depl.Name, err)
+	}
+	r.log.Info("created externalDNS deployment", "namespace", depl.Namespace, "name", depl.Name)
+	return nil
+}
+
+// updateExternalDNSDeployment updates the in-cluster externalDNS deployment.
+// Returns a boolean if an update was made, and an error when relevant.
+func (r *reconciler) updateExternalDNSDeployment(ctx context.Context, current, desired *appsv1.Deployment) (bool, error) {
+	// don't always update (or do simple DeepEqual with) the operand's deployment
+	// as this may result into a "fight" between API/admission webhooks and this controller
+	// example:
+	//  - this controller creates a deployment with the desired fields A, B, C
+	//  - API adds some default fields D, E, F (i.e. metadata, imagePullPullPolicy, dnsPolicy)
+	//  - mutating webhooks add default values to fields E, F
+	//  - this controller gets into reconcile loop and starts all from step 1
+	//  - checking that fields A, B, C are the same as desired would save us the before mentioned round trips:
+	changed, updated := externalDNSDeploymentChanged(current, desired)
+	if !changed {
+		return false, nil
+	}
+
+	if err := r.client.Update(ctx, updated); err != nil {
+		return false, fmt.Errorf("failed to update externalDNS deployment %s/%s: %w", desired.Namespace, desired.Name, err)
+	}
+	r.log.Info("updated externalDNS deployment", "namespace", desired.Namespace, "name", desired.Name)
+	return true, nil
+}
+
+// externalDNSDeploymentChange evaluates whether or not a deployment update is necessary.
+// Returns a boolean if an update is necessary, and the deployment resource to update to.
+func externalDNSDeploymentChanged(current, expected *appsv1.Deployment) (bool, *appsv1.Deployment) {
+	changed := false
+	updated := current.DeepCopy()
+
+	if len(current.Spec.Template.Spec.Containers) > 0 && len(expected.Spec.Template.Spec.Containers) > 0 {
+		if current.Spec.Template.Spec.Containers[0].Image != expected.Spec.Template.Spec.Containers[0].Image {
+			updated.Spec.Template.Spec.Containers[0].Image = expected.Spec.Template.Spec.Containers[0].Image
+			changed = true
+		}
+		currArgs := append([]string{}, current.Spec.Template.Spec.Containers[0].Args...)
+		expArgs := append([]string{}, expected.Spec.Template.Spec.Containers[0].Args...)
+		sort.Strings(currArgs)
+		sort.Strings(expArgs)
+		if !cmp.Equal(currArgs, expArgs) {
+			updated.Spec.Template.Spec.Containers[0].Args = expected.Spec.Template.Spec.Containers[0].Args
+			changed = true
+		}
+	}
```

Comment:
We want to be as impermissive of changes to operator-managed resources as possible—stomp any and all changes by users, but allow for defaulting behavior by the API and mutating admission plugins.  We can tighten the update logic up in a follow-up, but once people start using the operator, they may start doing all sorts of weird and unanticipated hacks if we allow it, and then it becomes much more difficult to tighten the operator's control over its resources.

Reply:
Do you have in mind some concrete fields we can additionally check?

Reply:
Updated the change detection logic and unit tests. Hashing of the zone is added too (needed for the change detecction).
Put everything in a separate commit, just in case: https://github.com/openshift/external-dns-operator/pull/19/commits/c0c6c2b55751ad45c310a3872a138ad941cc8aa6

Reply:
As discussed in the chat, the last implementation from https://github.com/openshift/external-dns-operator/commit/7ac3c115f2696eb94d4ed5f1298374a19eb6c20d looks fine to begin with.